raft/src/main/java/org/apache/kafka/raft/DynamicVoter.java

```diff
@@ -0,0 +1,188 @@
+    public static DynamicVoter parse(String input) {
+        input = input.trim();
+        int atIndex = input.indexOf("@");
+        if (atIndex < 0) {
+            throw new IllegalArgumentException("No @ found in dynamic voter string.");
+        }
+        if (atIndex == 0) {
+            throw new IllegalArgumentException("Invalid @ at beginning of dynamic voter string.");
+        }
+        String idString = input.substring(0, atIndex);
+        int nodeId;
+        try {
+            nodeId = Integer.parseInt(idString);
+        } catch (NumberFormatException e) {
+            throw new IllegalArgumentException("Failed to parse node id in dynamic voter string.", e);
+        }
+        if (nodeId < 0) {
+            throw new IllegalArgumentException("Invalid negative node id " + nodeId +
+                " in dynamic voter string.");
+        }
+        input = input.substring(atIndex + 1);
+        if (input.isEmpty()) {
+            throw new IllegalArgumentException("No hostname found after node id.");
+        }
+        String host;
+        if (input.startsWith("[")) {
+            int endBracketIndex = input.indexOf("]");
+            if (endBracketIndex < 0) {
+                throw new IllegalArgumentException("Hostname began with left bracket, but no right " +
+                        "bracket was found.");
+            }
+            host = input.substring(1, endBracketIndex);
+            input = input.substring(endBracketIndex + 1);
+        } else {
+            int endColonIndex = input.indexOf(":");
+            if (endColonIndex < 0) {
+                throw new IllegalArgumentException("No colon following hostname could be found.");
+            }
+            host = input.substring(0, endColonIndex);
+            input = input.substring(endColonIndex);
+        }
+        if (!input.startsWith(":")) {
+            throw new IllegalArgumentException("Port section must start with a colon.");
+        }
+        input = input.substring(1);
+        int endColonIndex = input.indexOf(":");
+        if (endColonIndex < 0) {
+            throw new IllegalArgumentException("No colon following port could be found.");
+        }
+        String portString = input.substring(0, endColonIndex);
+        int port;
+        try {
+            port = Integer.parseInt(portString);
+        } catch (NumberFormatException e) {
+            throw new IllegalArgumentException("Failed to parse port in dynamic voter string.", e);
+        }
+        if (port < 0 || port > 65535) {
+            throw new IllegalArgumentException("Invalid port " + port + " in dynamic voter string.");
+        }
+        String directoryIdString = input.substring(endColonIndex + 1);
+        Uuid directoryId;
+        try {
+            directoryId = Uuid.fromString(directoryIdString);
+        } catch (IllegalArgumentException e) {
+            throw new IllegalArgumentException("Failed to parse directory ID in dynamic voter string.", e);
+        }
```

Comment:
I did it this way because having a UUID in the middle of the whole thing kills readability for me.

When I look at:
`0@localhost:8082:JEXY6aqzQY-32P5TStzaFg`

I can first see

`0@localhost:8082`

Which makes sense and is familiar. And then I see the stuff at the end.

When I look at:

`0-JEXY6aqzQY-32P5TStzaFg@localhost:8082`

It's not visually clear whether the UUID is `0-JEXY6aqzQY-32P5TStzaFg` or `-JEXY6aqzQY-32P5TStzaFg`, or what.

Reply:
I see. You convinced me. replica id, host and port are more human readable so maybe they should come first in the string. In summary, we are going to keep the current schema of `<id>@<host>:<port>:<directoryId>`. Can you update the description for `--initial-voters` to match the implemented schema?

Reply:
Updated